### PR TITLE
sampling: introduce event processor

### PR DESCRIPTION
**Pull request by axw**
## Motivation/summary

Implement a tail-based sampling event processor.

The processor passes root transactions through a reservoir sampler, and periodically "finalizes" the reservoirs to make local sampling decisions. The resulting sampled trace events are then indexed into Elasticsearch; sampled trace IDs are published for other APM Servers to index related events.

To be done in follow-ups:
 - introduce config and wire up the processor in x-pack/apm-server/main.go
 - add system tests

## Checklist

- [x] I have signed the [Contributor License Agreement](https://www.elastic.co/contributor-agreement/).
~- [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/master/CHANGELOG.asciidoc)~

I have considered changes for:
~- [ ] documentation~
- [x] logging (add log lines, choose appropriate log selector, etc.)
~- [ ] metrics and monitoring (create issue for Kibana team to add metrics to visualizations, e.g. [Kibana#44001](https://github.com/elastic/kibana/issues/44001))~
- [x] automated tests (add tests for the code changes, all [**unit** tests](https://github.com/elastic/apm-server/blob/master/TESTING.md) pass locally)
~- [ ] telemetry~
~- [ ] Elasticsearch Service (https://cloud.elastic.co)~
~- [ ] Elastic Cloud Enterprise (https://www.elastic.co/products/ece)~
~- [ ] Elastic Cloud on Kubernetes (https://www.elastic.co/elastic-cloud-kubernetes)~

## How to test these changes

`make test`

## Related issues

Part of elastic/apm-server#4185

### CODE REVIEW

File: x-pack/apm-server/sampling/processor.go [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+		db:       db,
+		storage:  readWriter,
+		stopping: make(chan struct{}),
+		stopped:  make(chan struct{}),
+	}
+	return p, nil
+}
+
+// ProcessTransformables processes events, writing head sampled transactions and
+// spans to storage (except where they are part of a trace that has already been
```
Comment (simitt) [inline x-pack/apm-server/sampling/processor.go:79]: >  writing head sampled transactions and spans to storage (except where they are part of a trace that has already been tail sampled)

a bit confused by this - only head-based unsampled transactions are passed through. Maybe adopt the comment here to be more clear on what happens to events that have head-based sampling information.

File: x-pack/apm-server/sampling/processor.go [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+		// for a sampling decision.
+		return true, p.storage.WriteTransaction(tx)
+	}
+
+	// Root transaction: apply reservoir sampling.
+	reservoirSampled, err := p.groups.sampleTrace(tx)
+	if err == errTooManyTraceGroups {
+		// Too many trace groups, drop the transaction.
+		//
+		// TODO(axw) log a warning with a rate limit.
```
Comment (simitt) [inline x-pack/apm-server/sampling/processor.go:143]: +1 on logging

File: x-pack/apm-server/sampling/processor.go [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+	errgroup, ctx := errgroup.WithContext(context.Background())
+	errgroup.Go(func() error {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-p.stopping:
+			return context.Canceled
+		}
+	})
+	errgroup.Go(func() error {
```
Comment (simitt) [inline x-pack/apm-server/sampling/processor.go:280]: Could you add a comment to this `errgroup` please

**Review by simitt — APPROVED** [review]
only minor comments

File: x-pack/apm-server/sampling/processor.go (hunk 1 above)
Comment (axw) [reply x-pack/apm-server/sampling/processor.go:79]: We also pass through events for which we've already made/received a tail-sampling decision. i.e. when p.storage.IsTraceSampled returns no error. It's definitely a confusing message though, I'll try and clarify.

**Comment by simitt** [comment]
nothing to be tested here manually as the processor is not used yet.

**Pull request merged**